main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1316,31 +1347,48 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   /**
     * Returns the implicit type parameters constructed from the given formal parameters and return type.
     */
-  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): List[NamedAst.TypeParam] = {
+  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
     // Compute the type variables that occur in the formal parameters.
-    val typeVarsArgs = fparams.foldLeft(Set.empty[String]) {
-      case (acc, WeededAst.FormalParam(_, _, None, _)) => acc
-      case (acc, WeededAst.FormalParam(_, _, Some(tpe), _)) =>
-        freeVars(tpe).foldLeft(acc) {
-          case (innerAcc, ident) => innerAcc + ident.name
-        }
+    val typeVarsArgs = fparams.flatMap {
+      case WeededAst.FormalParam(_, _, Some(tpe), _) => freeVarsWithKind(tpe)
+      case WeededAst.FormalParam(_, _, None, _) => List.empty
     }
 
     // Compute the type variables that occur in the return type.
-    val typeVarsReturnType = freeVars(returnType).map(_.name)
+    val typeVarsReturnType = freeVarsWithKind(returnType)
 
     // Compute the set of type variables.
-    val typeVars = typeVarsArgs ++ typeVarsReturnType
+    val typeVars = typeVarsReturnType ::: typeVarsArgs
 
-    // Construct a (sorted) list of type parameters.
-    typeVars.toList.sorted.map {
-      case name =>
-        val ident = Name.Ident(SourcePosition.Unknown, name, SourcePosition.Unknown)
-        // We use a kind variable since we do not know the kind of the type variable.
-        val tvar = Type.freshVar(Kind.freshVar())
-        tvar.setText(name)
-        NamedAst.TypeParam(ident, tvar, loc)
+
+    val kindPerName = typeVars.groupBy(_._1.name).map {
+      case (name, kinds) => checkKinds(name, kinds)
+    }
+
```

Comment:
I have a suggestion for a better structure:

You have typeVarsArgs, which is a list of pairs of (type var, kind) (if I am not mistaken? - dont have my IDE right here)

You really just want to fold these into a map. However, since the fold can fail, you want to use a validation.

So I think you can use Validation.fold 

(Although now I see that it has been marked for deprecating-- but I don't remember why).

Am I making sense?

Reply:
I think I understood; I've updated the section using `Validation.foldRight` since that method isn't deprecated.